feat: custom single option select component [WD-10960]

mas-who opened the pull request:
## Done

- Created a custom single option select component that is searchable and can have its option labels modified.

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - Navigate to Permissions -> Groups
    - try modify permissions for a group, make sure the interaction with the permission selector is as expected.

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://lxd-ui-912.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/16119)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

mas-who commented:
@edlerd I updated the implementation for `PermissionSelector` with the new custom selector component. Haven't modified the option labels for the selectors yet since but have added `title` attributes to them which was previously not possible. I think we should discuss what's a good design for the permission selectors in terms of option labels and maybe do that as a follow up task? As an example, we can do something like what's shown below:
![Screenshot from 2024-09-18 17-05-26](https://github.com/user-attachments/assets/ca41e582-da36-4f66-b3d7-f8bb064601d2)

Lastly, I still need to fix the permission e2e tests, but thought we should have a discussion around about the above topic first since these tests could be affected.

edlerd commented:
> we can do something like what's shown below (this is not part of this PR at the moment):

Interesting idea, but we should not use a chip as non-interactive element as per the [design guidelines](https://vanillaframework.io/docs/patterns/chip/design-guidelines). They should be interactive, here we need a component that is informational, not interactive.

mas-who commented:
> > we can do something like what's shown below (this is not part of this PR at the moment):
> 
> Interesting idea, but we should not use a chip as non-interactive element as per the [design guidelines](https://vanillaframework.io/docs/patterns/chip/design-guidelines). They should be interactive, here we need a component that is informational, not interactive.

Aah thanks for the tip! This is currently just a demonstration for what we can do with the new select component :slightly_smiling_face: Not sure if we should update how the option labels as part of this PR though since it might broaden the scope a bit?

edlerd reviewed (commented):
The select works nicely and looks great!

One QA issue:
- when having many options with some being truncated, then searching for options that have short description, the dropdown width decreases for me

Also a couple of other ideas below.

mas-who commented on src/sass/_custom_select.scss:
```diff
@@ -0,0 +1,70 @@
+.p-select {
```
I specifically looked for this class in vanilla and it does not exist :sweat_smile: in vanilla, styling is applied directly to the `select` html element. But based on some of the other comments, if we change the component and other types to have "custom" in their names, then I think `p-custom-select` sounds good. When we get to upstreaming, some of these names might need to change though.

edlerd replied:
I think `CustomSelect` should be free in react-components as well, is it? Otherwise we should find a better name right away.

mas-who replied:
`CustomSelect` does not exist in react-components so we are good :slightly_smiling_face:

mas-who commented:
> The select works nicely and looks great!
> 
> One QA issue:
> 
> * when having many options with some being truncated, then searching for options that have short description, the dropdown width decreases for me

Thanks for catching this, previously I had applied a `min-width` for the dropdown container. This caused the behaviour you observed. Instead, the `width` property should be set which resolves the problem.

mas-who commented on src/util/permissions.tsx:
```diff
@@ -9,103 +8,108 @@ import {
   {
     value: "server",
     label: "Server",
+    text: "Server",
```
sounds good, added a util function `getOptionText` that implements the suggested logic above.

mas-who commented on src/util/permissions.tsx:
```diff
@@ -114,36 +118,28 @@ export const getResourceTypeOptions = (metadata?: LxdMetadata | null) => {
-  }
-
-  return (a.label ?? "").localeCompare(b.label as string);
+const sortOptions = (a: CustomOption, b: CustomOption): number => {
+  return (a.text ?? a.value).localeCompare(b.text ?? b.value);
```
I would keep this outside of the custom select component since sorting is not currently part of its responsibilities. Unless we want to include a sort option for the custom select?

edlerd reviewed (commented):
I think adding a drop shadow to the dropdown container and eventually a separator between the options might help the layout.

It doesn't look like a set of options to select from to me. I think the design is missing a 
stronger connection of the toggle to the dropdown. This is partly because the dropdown has a white background, it might improve with giving the same background to the options, that the toggle has now.

Hardcoding the width of the options dropdown is not ideal, before the same width of dropdown and toggle was making the connection between them very clear: 

![image](https://github.com/user-attachments/assets/55aa888d-9b5c-4ebe-a512-3b734a6c7fb3)

comparing to the new wider dropdown:

![image](https://github.com/user-attachments/assets/5c785f9a-1a98-4c36-8cc6-4428a81e6db0)

mas-who commented:
> I think adding a drop shadow to the dropdown container and eventually a separator between the options might help the layout.
> 
> It doesn't look like a set of options to select from to me. I think the design is missing a stronger connection of the toggle to the dropdown. This is partly because the dropdown has a white background, it might improve with giving the same background to the options, that the toggle has now.

Yes agreed, looking at the current `Select` component, the toggle and options have `background-color` set to `--vf-color-background-inputs` and `--vf-color-background-alt` respectively. Furthermore, the dropdown menu has a border with more visible drop shadow as you mentioned, I think applying all of those styling should make the select toggle more "connected" to the options.

> Hardcoding the width of the options dropdown is not ideal, before the same width of dropdown and toggle was making the connection between them very clear:
> 
> ![image](https://private-user-images.githubusercontent.com/1155472/370022558-55aa888d-9b5c-4ebe-a512-3b734a6c7fb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNjMzNDQsIm5iZiI6MTcyNzE2MzA0NCwicGF0aCI6Ii8xMTU1NDcyLzM3MDAyMjU1OC01NWFhODg4ZC05YjVjLTRlYmUtYTUxMi0zYjczNGE2YzdmYjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjRUMDczMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWFhMmIwYzQzMGEwNmRhNmQ0YjZkYWI0ZTY4NGQwOTUyNjQ4M2JmOTZmOWI0ZWE1MjJiOTk3NDQyZTMzNmZlNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.L2LhOVYHNetH5PoYgwuehFgLLKPGzQQcatetT-vUggU)
> 
> comparing to the new wider dropdown:
> 
> ![image](https://private-user-images.githubusercontent.com/1155472/370022855-5c785f9a-1a98-4c36-8cc6-4428a81e6db0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNjMzNDQsIm5iZiI6MTcyNzE2MzA0NCwicGF0aCI6Ii8xMTU1NDcyLzM3MDAyMjg1NS01Yzc4NWY5YS0xYTk4LTRjMzYtOGNjNi00NDI4YTgxZTZkYjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjRUMDczMDQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDExMDFjNWM5N2E1OTNmYTg5MmE3MDc1MTJhMGJiN2NmYzY3OGNlZjBmMGQyZGViNWMzYWRjYzE5MDY2NzQ5NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T4xiNqGIvJyTHvagIn0Uz5ga_GxJHkvQ29dK4ThZGqY)

If we want the dropdown to be the same width as the toggle by default, doing that through pure css is not possible as far as I can tell since the toggle is not a parent for the dropdown so we can't constrain its width. If we want it as a default behaviour in the custom select component, I think we will need to use JS. Alternatively, a user of the component can also style both the toggle and the dropdown accordingly. Maybe a combination of both makes sense?

mas-who commented:
@edlerd see below an attempt to style the custom select more in alignment with the current select component. Also noted that the toggle needed an active outline. Let me know what you think?

![Screenshot from 2024-09-24 10-41-05](https://github.com/user-attachments/assets/2a7b7a8a-c3bb-4bc7-89b0-16899508b18b)

Can we also maybe have a quick call to align on how to set the width for the dropdown menu when you have some time?

mas-who commented:
@edlerd as per our discussion from yesterday, I've made the following updates:
1. Removed the border for the dropdown menu
2. Increased the default max-height for the dropdown menu
3. implemented the logic to adjust the dropdown menu height to be always within the viewport height
Let me know what you think?

edlerd commented:
Thanks for the updates, it looks great :)

I think the dynamic height needs a bit refinement.

- When reducing the height of the browser, the dropdown shrinks, but when increasing, it keeps the old height
- When the dropdown is facing up, the height calculation seems to be a bit off:

![image](https://github.com/user-attachments/assets/5b813fd3-1ee0-4519-96c0-d1915649de33)

When reducing the height a bit more the size grows again:

![image](https://github.com/user-attachments/assets/7d21a59e-e7d0-4544-8e0f-f12a88205b01)

mas-who commented:
> hanks for the updates, it looks great :)

Thanks @edlerd I've added some updates. The way we adjust the height when the dropdown is below or above the toggle should be different, good catch there. The dropdown should also now adjust to the viewport height when increasing.

edlerd commented:
Also: why don't we remove the search bar on lists with <5 items any more? I think that was working fine before.

mas-who commented:
> Also: why don't we remove the search bar on lists with <5 items any more? I think that was working fine before.

Oops, I saw some permissions e2e tests failing since I was using the search to find the relevant option. I adjusted the search option to see if that was the case, but forgot to revert the changes and also fix the test :sweat_smile:

mas-who commented:
@edlerd I've reduced the option padding to 0.25rem and also reverted the permission selectors to auto show search now

edlerd commented:
Thanks, looks better with the reduced above/blow spacing. Maybe keep left/right padding at 0.5rem, I think that looked better. 

There is some spacing below the last and above the first item in the dropdown. A white stripe is visible, when hovering the first/last option in the dropdown. The first item only in cases where the search input is hidden.

mas-who commented:
> Thanks, looks better with the reduced above/blow spacing. Maybe keep left/right padding at 0.5rem, I think that looked better.
> 
> There is some spacing below the last and above the first item in the dropdown. A white stripe is visible, when hovering the first/last option in the dropdown. The first item only in cases where the search input is hidden.

@edlerd I've made adjustments based on the above comments, let me know what you think?

edlerd reviewed (commented):
QA looks mostly good now, found one more issue below.
Several comments for the code.

mas-who commented on src/components/select/CustomSelect.tsx:
```diff
@@ -0,0 +1,259 @@
+};
+
+const adjustDropdownHeight = () => {
+  const dropdown = document.querySelector(
+    ".p-custom-select__dropdown .p-list",
+  ) as HTMLElement;
+  const dropdownRect = dropdown.getBoundingClientRect();
```
I can't seem to recreate the error on the demo server or locally (using device view or just adjusting the height of the browser), I wonder if it's a browser version issue? Maybe you can show me how you did it?

I've updated the code to use refs as suggested, but only for the dropdown list and search input. The actual toggle is inside the `ContextualMenu` component which means an upstream change will be required for the `ContextualMenu` and `Button` components (since the ref must be attached to the button). Alternatively, I can pass a toggle `ReactNode` into `ContextualMenu`, but that's ugly as I would copy the button code and move it into the `CustomSelect` component. I think to limit the scope of this PR, a `querySelector` for the toggle element should be okay?

edlerd replied:
Can't reproduce it any more either, seems resolved.

mas-who commented on src/components/select/CustomSelect.tsx:
```diff
@@ -0,0 +1,259 @@
+  dropdown.style.maxHeight = `${adjustedHeight}px`;
+};
+
+const adjustDropdownHeightAbove = (dropdown: HTMLElement) => {
```
This should be resolved now, let me know if it's okay on your end as well?

![Screenshot from 2024-09-28 10-27-29](https://github.com/user-attachments/assets/9d4307f2-d4a2-4d0e-a263-ae8e895dc7f1)

mas-who commented on src/components/select/CustomSelect.tsx:
```diff
@@ -0,0 +1,259 @@
+  const adjustedHeight = bottomToViewportTop - searchHeight - 20;
+  dropdown.style.height = `${adjustedHeight}px`;
+  dropdown.style.maxHeight = `${adjustedHeight}px`;
+};
```
I've moved the functions into a `customSelect.tsx` utils file and created some test. Also I've cleaned up the functions a bit  to make it more readable. Although I still prefer having the above functions separate since a lot of the logical comparison as well as the offset calculations depend on the dropdown position. Creating a single function would mean we have to check the position of the dropdown and then decide on how to do the calculations, I think it would make the code more difficult to understand.

edlerd replied:
LGTM as is with the dedicated file and test applied.

mas-who commented on src/components/select/CustomSelectDropdown.tsx:
```diff
@@ -0,0 +1,196 @@
+} from "react";
+import classnames from "classnames";
+
+export type CustomSelectOption = LiHTMLAttributes<HTMLLIElement> & {
```
aah yes, we should destruct `option` into the `li` elements. Also I removed the `title` field from the `CustomSelectOption` type since it's already part of `LiHTMLAttributes`

mas-who commented on src/components/select/CustomSelectDropdown.tsx:
```diff
@@ -0,0 +1,196 @@
+
+  // filter options based on search text
+  if (search) {
+    options = options?.filter((option) => {
```
added the filter to the map render for `optionItems`

mas-who commented on tests/helpers/permission-groups.ts:
```diff
@@ -37,17 +37,26 @@ export const openEditGroupPanel = async (page: Page, groupName: string) => {
+  value: string,
+) => {
+  await page.getByRole("button", { name: label, exact: true }).click();
+  await page.getByRole("option", { name: value }).first().click();
```
right, needed to add `exact` to the selector

edlerd commented on src/sass/_custom_select.scss:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  // custom style introduced by this project
+  .p-custom-select__label {
```
The name here is confusing. This is about the toggleLabel, which is the text in the toggle. Not the label of the custom select itself.

edlerd commented on src/components/select/CustomSelect.tsx:
```diff
@@ -0,0 +1,163 @@
+  const searchRef = useRef<HTMLInputElement>(null);
+  const dropdownRef = useRef<HTMLUListElement>(null);
+
+  useEffect(() => {
+    if (takeFocus) {
+      const toggleButton = document.getElementById(selectId);
+      toggleButton?.focus();
+    }
+  }, [takeFocus]);
```
It works, when use the keyboard navigation, but with mouse the blue outline is missing.

mas-who replied:
thanks! Looking into it now

mas-who replied:
@edlerd the above issue is resolved, in addition I also noticed the following issues which are now fixed:
1. The edit permission panel was always focusing on the entitlement selector on first render. This was a pre-existing issue actually, now we will always focus on the resource type selector on first render.
2. When a select dropdown is open, keyboard event was not triggering for a dropdown without the search input. This is because the custom select uses a `div` for the dropdown container, which is not by default focusable and therefore cannot capture keyboard events. This is resolved by adding tabindex to the `div` container and focusing on it if the select is not searchable.

edlerd reviewed (commented):
Found one more issue on QA, code also seems all right, just one naming issue that confused me.

edlerd reviewed (approved):
QA and code LGTM, one idea to simplify a bit more.

mas-who commented on src/components/select/CustomSelect.tsx:
```diff
@@ -0,0 +1,163 @@
+    error?: ReactNode;
+    // Help text to show below the field.
+    help?: ReactNode;
+    // Styling for the wrapping Field component
```
yes agreed and done, just need one more approval :pray:
